### ISSUE TITLE
Strings with forced UTF-8 encoding trip headers table encoder

### ISSUE THREAD

**Issue by michaelklishin** [ruby-amqp/amq-protocol#30]
From a report on the mailing list:

``` ruby
#!/usr/bin/ruby
# encoding: utf-8
require "amqp"
imq = "Иван Иванов".force_encoding('utf-8')
AMQP.start(:host => "localhost") do |conn|
  ch = AMQP::Channel.new(conn)
  ex = ch.topic("")
  ex.publish(
    'Проверка проверяния проверки'.force_encoding('UTF-8'),
    :routing_key=>'hsm.CreateUser'.force_encoding('UTF-8'),
    :persistent => true,
    :headers=>{
      :id=>12321321,
      :login=>'ivano'.force_encoding('UTF-8'),
      :email=>'h@h'.force_encoding('UTF-8'),
      :phone=>'8055645654'.force_encoding('UTF-8'),
      :name=>imq}) do
    conn.close { EventMachine.stop } 
  end
end
```

As you see all the strings are static and forced to utf-8.

The question is: should the result below be considered as a bug, or a future?

The result:

```
$ ./sender.rb 
/home/me/.gem/ruby/1.9.1/gems/amq-protocol-1.2.0/lib/amq/protocol/table.rb:47:in `block in encode': incompatible character encodings: ASCII-8BIT and UTF-8 (Encoding::CompatibilityError)
from /home/me/.gem/ruby/1.9.1/gems/amq-protocol-1.2.0/lib/amq/protocol/table.rb:38:in `each'
  from /home/me/.gem/ruby/1.9.1/gems/amq-protocol-1.2.0/lib/amq/protocol/table.rb:38:in `encode'
  from /home/me/.gem/ruby/1.9.1/gems/amq-protocol-1.2.0/lib/amq/protocol/client.rb:1388:in `encode_headers'
  from /home/me/.gem/ruby/1.9.1/gems/amq-protocol-1.2.0/lib/amq/protocol/client.rb:1483:in `block in encode_properties'
```

**Comment by TomFreudenberg** [comment]
Hi Michael @michaelklishin

after years, I am running into nearly the same issue but only on `headers`

So writing

```ruby
Encoding.default_internal = Encoding::UTF_8
Encoding.default_external = Encoding::UTF_8

s = "äöüÄÖÜß"
puts s.encoding # => UTF-8

bunny_exchange.publish("payload", :persistent => true, :headers => { 'test' => s }, :routing_key => "test")
```

will raise

<pre>
gems/gems/amq-protocol-2.0.0/lib/amq/protocol/client.rb:1440:in `encode_properties': incompatible character encodings: ASCII-8BIT and UTF-8 (Encoding::CompatibilityError)
</pre>

If I save that as payload content, everything works fine

```ruby
bunny_exchange.publish(s, :persistent => true, :headers => { 'test' => "some content" }, :routing_key => "test")
```

Do you have any suggestion for me?

Thanks in advance for your feedback
Tom

**Comment by TomFreudenberg** [comment]
After some debugging, I was checked line where the error occurs

https://github.com/ruby-amqp/amq-protocol/blob/master/lib/amq/protocol/client.rb#L1440

If I check the `encoding` of `pieces.join()` it is correct as => UTF-8

The problem is based on the FIX-define of result in lines 

https://github.com/ruby-amqp/amq-protocol/blob/master/lib/amq/protocol/client.rb#L1437-L1439

Isn't it that you should force the encoding to result as the same encoding from pieces?

I fix it by using:

```ruby
pieces_as_string = pieces.join(EMPTY_STRING)
result.force_encoding(pieces_as_string.encoding) + pieces_as_string
```

If I interpret it correctly, `Array.join()` will check and set a compatible encoding to the resulting string.

Following tests were made:

```ruby
s1 = "abc".force_encoding("ASCII-8BIT")
s2 = "def".force_encoding("UTF-8")
s3 = "äöü" # auto UTF-8

a = [ s1, s2 ]
s = a.join(" ")
puts s.encoding # => ASCII-8BIT

a = [ s2, s1 ]
s = a.join(" ")
puts s.encoding # => UTF-8

a = [ s1, s2, s3 ]
s = a.join(" ")
puts s.encoding # => UTF-8
```

So as written above, I assume to use the fix to check encoding for the result before concatenate them.

If you like, I may send in a PR

Cheers
Tom

**Comment by michaelklishin** [comment]
Please do.

**Comment by TomFreudenberg** [comment]
Done ;-)

**Comment by TomFreudenberg** [comment]
Finished plus #57

**Comment by michaelklishin** [comment]
FTR, the fix in #57 was applied to a generated file, which was spotted over 1 year later and applied correctly as part of #62.
